File: java/ArraySpiral/ArraySpiral.java
import java.util.*;

public class ArraySpiral {

  private static enum Direction {
    UP,
    DOWN,
    LEFT,
    RIGHT;
  }

  public static List<List<Integer>> generateSpiralLists(int rowCount, int colCount) {
    // Initialize grid
    List<List<Integer>> grid = new ArrayList<>(rowCount);
    for (int i = 0; i < rowCount; i++) {
      List<Integer> row = new ArrayList<>(colCount);
      for (int j = 0; j < colCount; j++) {
	row.add(0);
      }
      grid.add(row);
    }

    int x = 0;
    int y = 0;
    int xUpBound = colCount - 1;
    int yUpBound = rowCount - 1;
    int xLowBound = 0;
    int yLowBound = 0;
    int currentNumber = 1;
    Direction direction = Direction.RIGHT;
    while (currentNumber <= rowCount * colCount) {
      grid.get(y).set(x, currentNumber);

      if (direction == Direction.RIGHT && x == xUpBound) {
	direction = Direction.DOWN;
	yLowBound++;
      } else if (direction == Direction.DOWN && y == yUpBound) {
	direction = Direction.LEFT;
	xUpBound--;
      } else if (direction == Direction.LEFT && x == xLowBound) {
	direction = Direction.UP;
	yUpBound--;
      } else if (direction == Direction.UP && y == yLowBound) {
	direction = Direction.RIGHT;
	xLowBound++;
      }
     
      switch (direction) {
	case RIGHT:
	  x++;
	  break;
	case DOWN:
	  y++;
	  break;
	case LEFT:
	  x--;
	  break;
	case UP:
	  y--;
	  break;
      }
      currentNumber++;
    }

    return grid;
  }

  public static void printListGrid(List<List<Integer>> grid) {
    System.out.println("- - - - - - - - - -");
    for (List<Integer> row : grid) {
      for (Integer value : row) {
	System.out.print(value + "  ");
      }
      System.out.println();
    }
    System.out.println("- - - - - - - - - -");
  }

  public static void main(String[] args) {
    printListGrid(generateSpiralLists(1, 1));
    printListGrid(generateSpiralLists(2, 2));
    printListGrid(generateSpiralLists(3, 3));
    printListGrid(generateSpiralLists(4, 4));
    printListGrid(generateSpiralLists(5, 5));
    printListGrid(generateSpiralLists(3, 4));
    printListGrid(generateSpiralLists(4, 3));
    printListGrid(generateSpiralLists(1, 2));
    printListGrid(generateSpiralLists(2, 1));
    printListGrid(generateSpiralLists(10, 5));
    printListGrid(generateSpiralLists(5, 10));
    printListGrid(generateSpiralLists(10, 4));
    printListGrid(generateSpiralLists(100, 11));
  }
}

File: ruby/array_spiral.rb
require 'pp'

def generate_spiral_lists(row_count, col_count)
  grid = row_count.times.map{ [] }
  x, y = 0, 0
  x_low_bound, y_low_bound = 0, 0
  x_up_bound, y_up_bound = col_count - 1, row_count - 1
  direction = :right

  (1..(row_count * col_count)).each do |current_number|
    grid[y][x] = current_number
    
    if direction == :right && x == x_up_bound
      direction = :down
      y_low_bound += 1
    elsif direction == :down && y == y_up_bound
      direction = :left
      x_up_bound -= 1
    elsif direction == :left && x == x_low_bound
      direction = :up
      y_up_bound -= 1
    elsif direction == :up && y == y_low_bound
      direction = :right
      x_low_bound += 1
    end

    case direction
    when :right
      x += 1
    when :down
      y += 1
    when :left
      x -= 1
    when :up
      y -= 1
    end
  end
  
  grid
end

pp generate_spiral_lists(1, 1)
pp generate_spiral_lists(2, 2)
pp generate_spiral_lists(3, 3)
pp generate_spiral_lists(4, 4)
pp generate_spiral_lists(5, 5)
pp generate_spiral_lists(3, 4)
pp generate_spiral_lists(4, 3)
pp generate_spiral_lists(1, 2)
pp generate_spiral_lists(2, 1)
pp generate_spiral_lists(10, 5)
pp generate_spiral_lists(5, 10)
pp generate_spiral_lists(10, 4)
binding.irb
